File: fw/hal.c
#include "hal.h"

// Some random amount of delay. The actual amount will depend on what
// the compiler does. The pragmas are here to make it slightly more
// predictable.
#pragma GCC push_options
#pragma GCC optimize("O0")
void delay()
{
    for(int i = 0; i < 100 * 1000; i++)
    {
        __asm__("nop");
    };
}
#pragma GCC pop_options

#define GPIO_BASE_ADDRESS ((int *)0x40000000)
#define BIT_MASK(n) (1 << (n))

bool get_button(void)
{
    volatile int *gpio = GPIO_BASE_ADDRESS;
    return *gpio & BIT_MASK(8);
}

void set_led_on(int index, bool led_on)
{
    volatile int *gpio = GPIO_BASE_ADDRESS;
    if(led_on)
    {
        *gpio |= BIT_MASK(index);
    }
    else
    {
        *gpio &= ~BIT_MASK(index);
    }
}

File: fw/rust/src/hal.rs
use core::arch::asm;

pub(crate) fn delay() {
    for _ in 0..10 {
        unsafe {
            asm!("nop");
        }
    }
}

const GPIO_BASE_ADDRESS: *mut u32 = 0x40000000 as *mut u32;

const fn bit_mask(i: u32) -> u32 {
    1 << i
}

pub(crate) fn get_button() -> bool {
    unsafe {
        return (GPIO_BASE_ADDRESS.read_volatile() & bit_mask(8)) != 0;
    }
}

pub(crate) fn set_led_on(index: u32, led_on: bool) {
    if led_on {
        unsafe {
            let value = GPIO_BASE_ADDRESS.read_volatile();
            GPIO_BASE_ADDRESS.write_volatile(value | bit_mask(index));
        }
    } else {
        unsafe {
            let value = GPIO_BASE_ADDRESS.read_volatile();
            GPIO_BASE_ADDRESS.write_volatile(value & !bit_mask(index));
        }
    }
}
